gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,362 @@
+import org.hypertrace.gateway.service.v1.baseline.BaselineEntitiesResponse;
+import org.hypertrace.gateway.service.v1.common.Value;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This service calculates baseline values for an Entity across time range. This converts both
+ * aggregate functions and time series functions in selection expression to time series data and
+ * generate baseline, lower bound and upper bound. It also changes Time series requests to get more
+ * data points for proper baseline calculation
+ */
+public class BaselineServiceImpl implements BaselineService {
+
+  private static final BaselineEntitiesRequestValidator baselineEntitiesRequestValidator =
+          new BaselineEntitiesRequestValidator();
+
+  protected static final long DAY_IN_MILLIS = 86400000L;
+
+  private final AttributeMetadataProvider attributeMetadataProvider;
+  private final BaselineServiceQueryParser baselineServiceQueryParser;
+  private final BaselineServiceQueryExecutor baselineServiceQueryExecutor;
+  private final EntityIdColumnsConfigs entityIdColumnsConfigs;
+
+  public BaselineServiceImpl(
+      AttributeMetadataProvider attributeMetadataProvider,
+      BaselineServiceQueryParser baselineServiceQueryParser,
+      BaselineServiceQueryExecutor baselineServiceQueryExecutor,
+      EntityIdColumnsConfigs entityIdColumnsConfigs) {
+    this.attributeMetadataProvider = attributeMetadataProvider;
+    this.baselineServiceQueryParser = baselineServiceQueryParser;
+    this.baselineServiceQueryExecutor = baselineServiceQueryExecutor;
+    this.entityIdColumnsConfigs = entityIdColumnsConfigs;
+  }
+
+  public BaselineEntitiesResponse getBaselineForEntities(
+      String tenantId,
+      BaselineEntitiesRequest originalRequest,
+      Map<String, String> requestHeaders) {
+    BaselineRequestContext requestContext =
+        getRequestContext(tenantId, requestHeaders, originalRequest);
+    Map<String, AttributeMetadata> attributeMetadataMap =
+            attributeMetadataProvider.getAttributesMetadata(requestContext, originalRequest.getEntityType());
+    baselineEntitiesRequestValidator.validate(originalRequest, attributeMetadataMap);
+    String timeColumn =
+        AttributeMetadataUtil.getTimestampAttributeId(
+            attributeMetadataProvider, requestContext, originalRequest.getEntityType());
+    Map<String, BaselineEntity> baselineEntityAggregatedMetricsMap = new HashMap<>();
+    Map<String, BaselineEntity> baselineEntityTimeSeriesMap = new HashMap<>();
+
+    if (originalRequest.getBaselineAggregateRequestCount() > 0) {
+      // Aggregated Functions data
+      Period aggTimePeriod =
+          getPeriod(originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      long periodSecs = getPeriodInSecs(aggTimePeriod);
```

Comment:
Why do we need a period for baselines for simple aggregations(non-time series)?

Reply:
Simple aggregations converted into time series data to calculate standard deviation which will be used to calculate lowerBound and upperBound.

Reply:
ok